### PR TITLE
[ALLUXIO-2035] Using singleton for configuration

### PR DESCRIPTION
**Pull request by jsimsa**
Addresses: https://alluxio.atlassian.net/browse/ALLUXIO-2035

### CODE REVIEW

File: core/client/src/main/java/alluxio/hadoop/ConfUtils.java [hunk 1]
```diff
@@ -84,6 +80,6 @@ public static Configuration loadFromHadoopConfiguration(
       }
     }
     LOG.info("Loading Alluxio properties from Hadoop configuration: {}", alluxioConfProperties);
-    return Configuration.fromMap(alluxioConfProperties);
+    Configuration.merge(alluxioConfProperties);
```
Comment (aaudiber) [inline core/client/src/main/java/alluxio/hadoop/ConfUtils.java]: I've created https://alluxio.atlassian.net/browse/ALLUXIO-2034 to track support for multiple clients with their own configuration. Could you add a TODO and link it here, and add a comment that we assume here there is only one client configuration in the JVM?

File: minicluster/pom.xml [hunk 2]
```diff
@@ -61,6 +61,12 @@
     <!-- Other projects' test-jars -->
     <dependency>
       <groupId>org.alluxio</groupId>
+      <artifactId>alluxio-core-common</artifactId>
```
Comment (aaudiber) [inline minicluster/pom.xml]: alphabetize (also I might conflict with you by adding the same test-jar)

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java [hunk 3]
```diff
@@ -77,35 +75,23 @@ public AbstractLocalAlluxioCluster(long workerCapacityBytes, int userBlockSize)
...
+    startMaster();
     waitForMasterReady();
-
-    startWorker(conf);
-
+    startWorker();
     waitForWorkerReady();
+
+    // Reset contexts so that they pick up the master and worker configuration.
+    resetContext();
```
Comment (aaudiber) [inline minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java]: should this be called resetContexts?

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java [hunk 4]
```diff
@@ -443,16 +422,9 @@ public void run() {
...
-   */
-  public Configuration getMasterConf() {
-    return mMasterConf;
-  }
-
-  /**
    * Resets contexts. This is called when the cluster is stopped.
    */
-  protected void resetContext() {}
+  protected abstract void resetContext();
```
Comment (aaudiber) [inline minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java]: could the implementation be here? I think the subclasses implement this identically

File: minicluster/src/main/java/alluxio/master/LocalAlluxioCluster.java [hunk 5]
```diff
@@ -97,51 +96,33 @@ public AlluxioWorker getWorker() {
...
     mMaster.start();
 
     // Update the test conf with actual RPC port.
-    testConf.set(Constants.MASTER_RPC_PORT, String.valueOf(getMasterPort()));
-
-    // We need to update client context with the most recent configuration so they know the correct
-    // port to connect to master.
-    ClientContext.getConf().merge(testConf);
-    ClientTestUtils.reinitializeClientContext();
+    Configuration.set(Constants.MASTER_RPC_PORT, String.valueOf(getMasterPort()));
```
Comment (aaudiber) [inline minicluster/src/main/java/alluxio/master/LocalAlluxioCluster.java]: Is it necessary to set this both here and in LocalAlluxioMaster line 65?

File: tests/src/test/java/alluxio/master/AlluxioMasterRestApiTest.java [hunk 6]
```diff
@@ -134,14 +134,8 @@ public void getWorkerInfoListTest() throws Exception {
...
 
-  private Configuration mockConfiguration() {
-    Configuration conf = PowerMockito.spy(MasterContext.getConf());
-    PowerMockito.spy(MasterContext.class);
-    Mockito.when(MasterContext.getConf()).thenReturn(conf);
-    return conf;
-  }
-
   @Test
+  @Ignore // TODO(jiri): re-enable
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/master/AlluxioMasterRestApiTest.java]: what's going on here?

File: core/server/src/main/java/alluxio/master/MasterContext.java [hunk 7]
```diff
@@ -11,58 +11,31 @@
...
   public static MasterSource getMasterSource() {
     return sMasterSource;
   }
 
   /**
-   * Resets the master context, for test only.
-   * TODO(binfan): consider a better way to mock test configuration
+   * Resets the master context.
    */
   public static void reset() {
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/master/MasterContext.java]: I think we can drop this reset method now, or at least use it very rarely (I don't think tests care about the master source).

File: core/server/src/main/java/alluxio/worker/WorkerContext.java [hunk 8]
```diff
@@ -65,29 +45,16 @@ public static WorkerNetAddress getNetAddress() {
...
   public static void setWorkerNetAddress(WorkerNetAddress netAddress) {
     sNetAddress = netAddress;
   }
 
   /**
-   * Resets the worker context, for test only.
-   * TODO(binfan): consider a better way to mock test configuration
+   * Resets the worker context.
    */
   public static void reset() {
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/WorkerContext.java]: I think we can drop this reset method now, or at least use it very rarely (I don't think tests care about the worker source).

File: minicluster/src/main/java/alluxio/master/MultiMasterLocalAlluxioCluster.java [hunk 9]
```diff
@@ -205,17 +195,24 @@ protected void startMaster(Configuration conf) throws IOException {
       }
     }
     // Use first master port
-    mMasterConf.set(Constants.MASTER_RPC_PORT, String.valueOf(getMaster().getRPCLocalPort()));
+    Configuration.set(Constants.MASTER_RPC_PORT, String.valueOf(getMaster().getRPCLocalPort()));
```
Comment (aaudiber) [inline minicluster/src/main/java/alluxio/master/MultiMasterLocalAlluxioCluster.java]: I think master.start sets this already

File: tests/src/test/java/alluxio/LocalAlluxioClusterResource.java [hunk 10]
```diff
@@ -130,36 +127,19 @@ public LocalAlluxioCluster get() {
...
-    applyConfParams();
-  }
-
   private void applyConfParams() throws IOException {
-    mTestConf = mLocalAlluxioCluster.newTestConf();
+    mLocalAlluxioCluster.initializeTestConfiguration();
     // Override the configuration parameters with mConfParams
     for (int i = 0; i < mConfParams.length; i += 2) {
-      mTestConf.set(mConfParams[i], mConfParams[i + 1]);
+      Configuration.set(mConfParams[i], mConfParams[i + 1]);
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/LocalAlluxioClusterResource.java]: should we reset the Configuration at the end?

File: tests/src/test/java/alluxio/client/concurrent/FileInStreamConcurrencyIntegrationTest.java [hunk 11]
```diff
@@ -35,8 +35,7 @@
  */
 public final class FileInStreamConcurrencyIntegrationTest {
   private static final int BLOCK_SIZE = 30;
-  private static final int READ_THREADS_NUM =
-      ClientContext.getConf().getInt(Constants.USER_BLOCK_MASTER_CLIENT_THREADS) * 10;
+  private static int sNumReadThreads;
```
Comment (aaudiber) [inline tests/src/test/java/alluxio/client/concurrent/FileInStreamConcurrencyIntegrationTest.java]: why is this initialized in a new way?

**Comment by aaudiber** [comment]
@jsimsa thanks for bringing sanity to our handling of configuration! I've left some comments on the code, but overall I think this is a significant improvement.

File: core/client/src/main/java/alluxio/hadoop/ConfUtils.java (hunk 1 above)
Comment (jsimsa) [inline core/client/src/main/java/alluxio/hadoop/ConfUtils.java]: done

File: core/server/src/main/java/alluxio/master/MasterContext.java (hunk 7 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/MasterContext.java]: It is invoked as part of ClientTestUtils.resetClient(), which is the recommended method for resetting all client state.

Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/MasterContext.java]: Disregard my previous comment. I will see if I can get rid of this.

File: core/server/src/main/java/alluxio/worker/WorkerContext.java (hunk 8 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/worker/WorkerContext.java]: done

File: minicluster/pom.xml (hunk 2 above)
Comment (jsimsa) [inline minicluster/pom.xml]: done

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java (hunk 3 above)
Comment (jsimsa) [inline minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java]: renamed to reset

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java (hunk 4 above)
Comment (jsimsa) [inline minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java]: done

File: core/server/src/main/java/alluxio/master/MasterContext.java (hunk 7 above)
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/MasterContext.java]: removed

File: tests/src/test/java/alluxio/client/concurrent/FileInStreamConcurrencyIntegrationTest.java (hunk 11 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/client/concurrent/FileInStreamConcurrencyIntegrationTest.java]: reverted

File: tests/src/test/java/alluxio/LocalAlluxioClusterResource.java (hunk 10 above)
Comment (jsimsa) [inline tests/src/test/java/alluxio/LocalAlluxioClusterResource.java]: when the cluster is stopped, configuration is reset

File: minicluster/src/main/java/alluxio/master/MultiMasterLocalAlluxioCluster.java (hunk 9 above)
Comment (jsimsa) [inline minicluster/src/main/java/alluxio/master/MultiMasterLocalAlluxioCluster.java]: we need to set it to the RPC port of the leader

File: minicluster/src/main/java/alluxio/master/LocalAlluxioCluster.java (hunk 5 above)
Comment (jsimsa) [inline minicluster/src/main/java/alluxio/master/LocalAlluxioCluster.java]: removed

File: tests/src/test/java/alluxio/LocalAlluxioClusterResource.java (hunk 10 above)
Comment (aaudiber) [inline tests/src/test/java/alluxio/LocalAlluxioClusterResource.java]: sg

File: minicluster/src/main/java/alluxio/master/MultiMasterLocalAlluxioCluster.java (hunk 9 above)
Comment (aaudiber) [inline minicluster/src/main/java/alluxio/master/MultiMasterLocalAlluxioCluster.java]: sg

**Comment by aaudiber** [comment]
LGTM

**Comment by gpang** [comment]
LGTM

**Comment by apc999** [comment]
LGTM

**Comment by pfxuan** [comment]
@jsimsa,  I have a question for this PR:

If we only keep a global `Configuration` across all workers, how do I maintain workers' individual state? Should I use `WorkerSource`?  Previously, I can use `WorkerContext.getConf()` to keep tracking the worker's states. 

Thanks!

**Comment by jsimsa** [comment]
@pfxuan what type of worker state would you like to keep track of?

**Comment by pfxuan** [comment]
@jsimsa, I want to control the size of in-memory storage space via dynamically tracking and adjusting the reserved ratio `alluxio.worker.tieredstore.level0.reserved.ratio` for each of workers. Do you know where is the best place to keep this local state?

**Comment by jsimsa** [comment]
if you do not mind modifying the source code, I would add this information to `WorkerContext` which you can think of as a singleton for worker state

**Comment by pfxuan** [comment]
This solution is good for me. Thanks for your suggestion!

**Pull request merged**